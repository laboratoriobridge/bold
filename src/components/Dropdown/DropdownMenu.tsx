import React from 'react'

import { useRovingTabIndex } from '../../hooks/useRovingTabIndex'
import { ExternalStyles, Theme, useStyles } from '../../styles'
import { Omit } from '../../util'
import { composeRefs } from '../../util/react'

export interface DropdownMenuProps extends Omit<React.HTMLAttributes<HTMLUListElement>, 'style'> {
  innerRef?: React.RefObject<HTMLUListElement>
  style?: ExternalStyles
}

export function DropdownMenu(props: DropdownMenuProps) {
  const { innerRef, style, ...rest } = props
  const { css, classes } = useStyles(styles)

  const rootRef = useRovingTabIndex({
    getItems: root => Array.from(root.querySelectorAll('[role="menuitem"]:not([aria-disabled="true"])')),
  })

  return <ul ref={composeRefs(innerRef, rootRef)} className={css(classes.root, style)} role='menu' {...rest} />
}

export const styles = (theme: Theme) => ({
  root: {
    listStyle: 'none',
    margin: 0,
    padding: 'calc(0.5rem - 1px) 0',
    whiteSpace: 'nowrap',
    boxShadow: '0 2px 4px 0 rgba(0, 0, 0, 0.09)',
    border: `1px solid ${theme.pallete.divider}`,
    borderRadius: theme.radius.popper,
    display: 'inline-block',
    width: 'auto',
    minWidth: '150px',
    background: theme.pallete.surface.main,
  } as React.CSSProperties,
})
