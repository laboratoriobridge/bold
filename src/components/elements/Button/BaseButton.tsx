import * as React from 'react'

export type OnClickWithPromise = (event: React.MouseEvent<any>) => any

export interface BaseButtonProps {
    /**
     * css className
     */
    className?: string
    disabled?: boolean
    name?: string
    onClick?: React.MouseEventHandler<any> | OnClickWithPromise
    onMouseEnter?: React.MouseEventHandler<any>
    onMouseLeave?: React.MouseEventHandler<any>
    tabIndex?: number
    title?: string
    innerRef?: React.Ref<HTMLElement>
    onLoadingChange?(loading: boolean): void
    render?(props: any): React.ReactNode
}

export class BaseButton extends React.Component<BaseButtonProps> {

    static defaultProps: BaseButtonProps = {
        render: ({ innerRef, ...rest }: any) => {
            return (
                <button type='button' ref={innerRef} {...rest} />
            )
        },
    }

    private timeout: number

    componentWillUnmount() {
        clearTimeout(this.timeout)
        this.timeout = -1
    }

    render() {
        const { onLoadingChange, render, ...rest } = this.props

        return render({
            ...rest,
            onClick: this.onClick,
            onKeyPress: this.handleOnKeyPress,
        })
    }

    private handleOnKeyPress = (event) => {
        event.preventDefault()
        if (!event) { event = window.event } // cross-browser shenanigans
        if (event.charCode === 32 || event.charCode === 13 && this.props.onClick) { // this is the spacebar
            this.onClick(event)
        }
        return true // treat all other keys normally;
    }

    private onClick = (event: React.MouseEvent<any>) => {
        if (this.props.onClick) {

            const promise = this.props.onClick(event)
            if (promise && promise.then) {
                this.startLoading()
                promise
                    .then(() => this.stopLoading())
                    .catch((error) => {
                        this.stopLoading()
                        throw new Error(error)
                    })
            }
        }
    }

    private startLoading = () => {
        this.props.onLoadingChange && this.props.onLoadingChange(true)
    }

    private stopLoading() {
        if (!this.timeout) {
            this.timeout = window.setTimeout(() => {
                this.props.onLoadingChange && this.props.onLoadingChange(false)
                this.timeout = null
            }, 10)
        }
    }

}
