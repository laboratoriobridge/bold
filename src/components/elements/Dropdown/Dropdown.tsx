import Downshift, { ControllerStateAndHelpers, DownshiftProps } from 'downshift'
import * as React from 'react'
import { Manager, Popper, Reference, ReferenceChildrenProps } from 'react-popper'

import { useTheme } from '../../../styles'
import { Portal } from '../Portal'
import { FadeTransition } from '../Transition/FadeTransition'

export interface DropdownProps extends DownshiftProps<any> {
    renderTarget(renderProps: DropdownTargetRenderProps): React.ReactNode
    children(renderProps: DropdownRenderProps): React.ReactNode
}

export type DropdownRenderProps = ControllerStateAndHelpers<any>
export type DropdownTargetRenderProps = ControllerStateAndHelpers<any> & ReferenceChildrenProps

export const Dropdown = (props: DropdownProps) => {
    const { children, renderTarget, ...rest } = props

    const itemToString = () => null
    const theme = useTheme()

    return (
        <Downshift itemToString={itemToString} suppressRefError {...rest}>
            {(downshift) => {
                return (
                    <Manager>
                        <Reference>
                            {(refProps) => renderTarget({ ...downshift, ...refProps })}
                        </Reference>
                        <FadeTransition in={downshift.isOpen}>
                            {({ className }) => (
                                downshift.isOpen && (
                                    <Portal>
                                        <Popper placement='bottom'>
                                            {({ ref, style, placement }) => (
                                                <div
                                                    ref={ref}
                                                    className={className}
                                                    style={{ ...style, zIndex: theme.zIndex.dropdown }}
                                                    data-placement={placement}
                                                >
                                                    {children(downshift)}
                                                </div>
                                            )}
                                        </Popper>
                                    </Portal>
                                )
                            )}
                        </FadeTransition>
                    </Manager>
                )
            }}
        </Downshift>
    )

}
