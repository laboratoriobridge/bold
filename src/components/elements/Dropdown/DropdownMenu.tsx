import * as React from 'react'

import { Styles, withStyles, WithStylesProps } from '../../../styles'
import { Tooltip } from '../Tooltip/Tooltip'

export interface DropdownMenuProps extends WithStylesProps {
    highlightedIndex?: number
}

@withStyles
export class DropdownMenu extends React.Component<DropdownMenuProps> {
    render() {
        const { css, theme } = this.props
        const styles: Styles = {
            list: {
                fontWeight: 'bold',
                color: theme.pallete.text.main,
                whiteSpace: 'nowrap',
                listStyle: 'none',
                margin: 0,
                padding: 0,
                boxShadow: '0 2px 4px 0 rgba(0, 0, 0, 0.09)',
                border: '1px solid ' + theme.pallete.divider,
                borderRadius: theme.radius.popper,
                display: 'inline-block',
                width: 'auto',
                minWidth: '150px',
                background: theme.pallete.surface.main,
                textAlign: 'left',
            },
        }

        return (
            <ul className={css(styles.list)}>
                {this.props.children}
            </ul>
        )
    }
}

export interface DropdownItemProps extends WithStylesProps {
    hint?: string
    disabled?: boolean
    type?: 'normal' | 'danger'
    highlighted?: boolean
    onSelected?(): any
}

@withStyles
export class DropdownItem extends React.Component<DropdownItemProps> {
    static defaultProps: DropdownItemProps = {
        disabled: false,
        type: 'normal',
        onSelected: () => null,
    }

    render() {
        const { css, theme, onSelected, type, disabled, hint, highlighted } = this.props
        const styles: Styles = {
            item: {
                margin: 0,
                'div': {
                    display: 'block', // override inline-block defined by tooltip wrapper
                },
                '&:first-of-type a': {
                    borderTopLeftRadius: theme.radius.popper,
                    borderTopRightRadius: theme.radius.popper,
                },
                '&:last-of-type a': {
                    borderBottomLeftRadius: theme.radius.popper,
                    borderBottomRightRadius: theme.radius.popper,
                },
            },
            link: {
                display: 'block',
                padding: '0.5rem 1rem',
                fontSize: theme.typography.sizes.button,
                color: theme.pallete.text.main,
                textDecoration: 'none',
                '&:hover': {
                    background: theme.pallete.surface.background,
                },
            },
            disabled: {
                color: theme.pallete.gray.c70,
                '&:hover': {
                    background: 'transparent',
                    cursor: 'not-allowed',
                },
            },
            danger: {
                color: theme.pallete.status.danger.main,
                '&:hover': {
                    color: theme.pallete.surface.main,
                    background: theme.pallete.status.danger.main,
                },
            },
            highlighted: {
                normal: {
                    background: theme.pallete.surface.background,
                },
                danger: {
                    color: theme.pallete.surface.main,
                    background: theme.pallete.status.danger.main,
                },
            },
        }
        const classes = css(styles.link,
            type === 'danger' && styles.danger,
            disabled && styles.disabled,
            highlighted && styles.highlighted[type]
        )

        const link = (
            <a onClick={disabled ? null : onSelected} className={classes}>
                {this.props.children}
            </a>
        )

        return (
            <li className={css(styles.item)}>
                {hint ? <Tooltip text={hint}>{link}</Tooltip> : link}
            </li>
        )
    }
}
