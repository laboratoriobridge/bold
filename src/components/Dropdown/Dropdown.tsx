import { Options as PopperOptions } from '@popperjs/core'
import React, { useCallback, useEffect, useState, useRef } from 'react'
import { usePopper } from 'react-popper'
import { useClickOutside } from '../../hooks/useClickOutside'
import { Theme, useStyles } from '../../styles'
import { randomStr } from '../../util/string'
import { DropdownMenu, DropdownMenuProps } from './DropdownMenu'

export interface DropdownProps extends DropdownMenuProps {
  /**
   * The anchor element that should receive the dropdown menu.
   */
  anchorRef: HTMLElement | null

  /**
   * Whether the dropdown menu is open or not.
   * When dropdown is opened, it automatically attaches aria attributes to anchor element and focus first menu item.
   */
  open: boolean

  /**
   * If true, the `onClose` callback is called when any menu item is clicked.
   */
  autoclose?: boolean

  /**
   * Popper instance props.
   */
  popperProps?: Partial<PopperOptions>

  children?: React.ReactNode

  /**
   * Called whenever the dropdown wants to close itself.
   *
   * It may be called if:
   * - `Escape` button is pressed
   * - Focus go outside dropdown menu
   * - Clicked anywhere outside anchor and dropdown menu
   * - Any item is clicked and `autoclose` prop is `true`
   */
  onClose?(): void
}

export function Dropdown(props: DropdownProps) {
  const { children, anchorRef, popperProps, open, onClose, autoclose, style, ...rest } = props
  const { classes } = useStyles(createStyles)

  const [menuRef, setMenuRef] = useState<HTMLUListElement>()

  const {
    styles: { popper: popperStyle },
    attributes: { placement },
  } = usePopper(anchorRef, menuRef, popperProps)

  const dropdownIdRef = React.useRef<string>(`dropdown-${randomStr()}`)

  // Attaches aria attributes to anchor element
  useEffect(() => {
    if (anchorRef) {
      anchorRef.setAttribute('aria-haspopup', 'true')

      if (open) {
        anchorRef.setAttribute('aria-expanded', 'true')
        anchorRef.setAttribute('aria-controls', dropdownIdRef.current)
      } else {
        anchorRef.removeAttribute('aria-expanded')
        anchorRef.removeAttribute('aria-controls')
      }
    }
  }, [open, anchorRef])

  // Attaches Escape key event
  useEffect(() => {
    const handleKeyDown = (e: KeyboardEvent) => {
      if (open && e.key === 'Escape') {
        onClose()
      }
    }

    document.addEventListener('keydown', handleKeyDown)
    return () => document.removeEventListener('keydown', handleKeyDown)
  }, [open, onClose])

  // When opened, focus the first menu item
  const isFirstRender = useRef(true)
  useEffect(() => {
    if (isFirstRender.current) {
      isFirstRender.current = false
      return
    }

    if (open) {
      setTimeout(() => {
        // Delay focus to preserve window scroll position
        if (menuRef) {
          const firstItem = menuRef.querySelector('[role="menuitem"]:not([aria-disabled="true"])') as HTMLElement

          if (firstItem) {
            firstItem.focus()
          }
        }
      })
    }
  }, [open, menuRef])

  // Call onClose if target focus is outside menu
  const handleBlur = useCallback(() => {
    setTimeout(() => {
      if (menuRef) {
        const currentFocus = menuRef.ownerDocument.activeElement
        if (!menuRef.contains(currentFocus)) {
          onClose()
        }
      }
    })
  }, [onClose, menuRef])

  // Call `onClose` when clicked outside dropdown and anchor
  useClickOutside([anchorRef, menuRef], () => {
    if (open) {
      onClose()
    }
  })

  const handleMenuClick = useCallback(() => {
    if (autoclose) {
      onClose()
    }
  }, [autoclose, onClose])

  return (
    open && (
      <DropdownMenu
        id={dropdownIdRef.current}
        innerRef={setMenuRef}
        style={[popperStyle as any, classes.dropdown, style]}
        data-placement={placement}
        onClick={handleMenuClick}
        onBlur={handleBlur}
        {...rest}
      >
        {children}
      </DropdownMenu>
    )
  )
}

Dropdown.defaultProps = {
  autoclose: true,
  onClose: () => null,
} as Partial<DropdownProps>

export const createStyles = (theme: Theme) => ({
  dropdown: {
    zIndex: theme.zIndex.dropdown,
  },
})
