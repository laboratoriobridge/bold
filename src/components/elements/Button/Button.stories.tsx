import { action } from '@storybook/addon-actions'
import { boolean, select, text } from '@storybook/addon-knobs'
import { storiesOf } from '@storybook/react'
import React from 'react'
import { MemoryRouter } from 'react-router'

import { Icon } from '../Icon'
import { Text } from '../textual'

import { Button } from './Button/Button'
import { ButtonKind, ButtonSize, ButtonSkin } from './Button/ButtonSkins'
import { ButtonLink } from './ButtonLink/ButtonLink'

const typeOptions: ButtonKind[] = ['normal', 'primary', 'danger']
const skinOptions: ButtonSkin[] = ['default', 'ghost', 'outline']
const sizeOptions: ButtonSize[] = ['small', 'medium', 'large']

storiesOf('Components|Button', module)
  .add('default', () => (
    <Button
      skin={select('skin', skinOptions, 'default')}
      size={select('size', sizeOptions, 'medium')}
      kind={select('kind', typeOptions, 'normal')}
      loading={boolean('loading', false)}
      disabled={boolean('disabled', false)}
      block={boolean('block', false)}
      onClick={action('button-clicked')}
    >
      {text('children', 'Button')}
    </Button>
  ))
  .add('icon', () => (
    <Button
      skin={select('skin', skinOptions, 'ghost')}
      size={select('size', sizeOptions, 'small')}
      kind={select('kind', typeOptions, 'normal')}
      loading={boolean('loading', false)}
      disabled={boolean('disabled', false)}
      block={boolean('block', false)}
      onClick={action('button-clicked')}
    >
      <Icon icon='adjust' />
    </Button>
  ))
  .add('link', () => (
    <MemoryRouter initialEntries={['/']} initialIndex={0}>
      <ButtonLink
        to='/links'
        skin={select('skin', skinOptions, 'default')}
        size={select('size', sizeOptions, 'medium')}
        kind={select('kind', typeOptions, 'normal')}
        loading={boolean('loading', false)}
        disabled={boolean('disabled', false)}
        block={boolean('block', false)}
        onClick={action('link-clicked')}
      >
        <Icon icon='arrowLeft' style={{ marginRight: '0.5rem' }} />
        <Text>Go back</Text>
      </ButtonLink>
    </MemoryRouter>
  ))
