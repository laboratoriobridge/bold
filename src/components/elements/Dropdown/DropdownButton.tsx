import * as React from 'react'

import { Omit } from '../../../util/types'
import { Button, ButtonProps } from '../Button'

import { Dropdown, DropdownTargetRenderProps } from './Dropdown'
import { DropdownItem, DropdownItemProps, DropdownMenu } from './DropdownMenu'

export type DropdownItemConfig = Omit<DropdownButtonItemProps, 'closeMenu'>

export interface DropdownButtonProps extends ButtonProps {
    items: DropdownItemConfig[]
}

export class DropdownButton extends React.PureComponent<DropdownButtonProps> {

    render() {
        const { items } = this.props
        return (
            <Dropdown renderTarget={this.renderButton} onChange={this.handleChange}>
                {({ closeMenu, highlightedIndex, getMenuProps, getItemProps }) => (
                    <DropdownMenu highlightedIndex={highlightedIndex} {...getMenuProps()}>
                        {items.map((item, idx) => (
                            <DropdownButtonItem
                                key={idx}
                                closeMenu={closeMenu}
                                {...getItemProps({ item })}
                                highlighted={highlightedIndex === idx}
                                {...item}
                            />
                        ))}
                    </DropdownMenu>
                )}
            </Dropdown>
        )
    }

    renderButton = ({ ref, getRootProps, getToggleButtonProps }: DropdownTargetRenderProps) => {
        const { items, ...other } = this.props
        return (
            <Button
                innerRef={ref}
                {...getRootProps()}
                {...getToggleButtonProps()}
                {...other}
            />
        )
    }

    handleChange = (item: DropdownItemConfig) => {
        item.onSelected && item.onSelected()
    }
}

export interface DropdownButtonItemProps extends DropdownItemProps {
    content: React.ReactNode
    autoClose?: boolean
    closeMenu?(): void
}

export class DropdownButtonItem extends React.PureComponent<DropdownButtonItemProps> {

    static defaultProps: Partial<DropdownButtonItemProps> = {
        content: '',
        autoClose: true,
    }

    render() {
        const { content, closeMenu, autoClose, ...other } = this.props
        return <DropdownItem {...other} onSelected={this.handleClick}>{content}</DropdownItem>
    }

    handleClick = () => {
        const { onSelected, autoClose, closeMenu } = this.props
        if (autoClose) {
            closeMenu()
        }
        onSelected && onSelected()
    }

}
