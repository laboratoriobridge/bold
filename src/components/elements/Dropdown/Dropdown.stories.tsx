import { action } from '@storybook/addon-actions'
import { storiesOf } from '@storybook/react'
import React from 'react'

import { Tooltip } from '../Tooltip'

import { DropdownButton } from './DropdownButton'
import { DropdownDivider } from './DropdownDivider'
import { DropdownItem } from './DropdownItem'
import { DropdownMenu } from './DropdownMenu'

storiesOf('Components|Dropdown', module)
  .add('button', () => (
    <DropdownButton
      size='small'
      items={[
        { content: 'Item #1', onClick: action('item 1 clicked') },
        { content: 'Item #2', onClick: action('item 2 clicked'), type: 'danger' },
        {
          content: 'Item #3',
          onClick: action('item 3 clicked'),
          disabled: true,
          tooltip: 'This is disabled',
        },
      ]}
    >
      Options
    </DropdownButton>
  ))
  .add('menu', () => (
    <DropdownMenu>
      <DropdownItem onClick={action('clicked item 1')}>Item 1</DropdownItem>
      <DropdownItem onClick={action('clicked item 2')} disabled>
        Item 2
      </DropdownItem>
      <DropdownItem onClick={action('clicked item 3')}>Item 3</DropdownItem>
      <DropdownDivider />
      <DropdownItem onClick={action('clicked item 4')} type='danger'>
        <Tooltip text='Some danger option'>
          <div>Item 4</div>
        </Tooltip>
      </DropdownItem>
    </DropdownMenu>
  ))
