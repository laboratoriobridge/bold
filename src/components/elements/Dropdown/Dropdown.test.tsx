import * as React from 'react'
import { fireEvent, render } from 'react-testing-library'

import { withTheme } from '../../../test'
import { Button } from '../Button'

import { Dropdown } from './Dropdown'
import { DropdownItem, DropdownMenu } from './DropdownMenu'

// tslint:disable jsx-no-lambda

const createDropdown = () => (
    <Dropdown
        renderTarget={({ ref, getRootProps, getToggleButtonProps }) =>
            <Button
                label='Menu'
                innerRef={ref}
                {...getRootProps()}
                {...getToggleButtonProps()}
            />}
    >
        {({ highlightedIndex }) => (
            <DropdownMenu highlightedIndex={highlightedIndex}>
                <DropdownItem>Item #1</DropdownItem>
                <DropdownItem>Item #2</DropdownItem>
            </DropdownMenu>
        )}
    </Dropdown>
)

it('should render correctly', () => {
    render(withTheme(createDropdown()))

    expect(document.body).toMatchSnapshot()
})

it('should open menu on click', () => {
    const { getByText } = render(withTheme(createDropdown()))

    fireEvent.click(getByText('Menu'))

    expect(document.body).toMatchSnapshot()

})
