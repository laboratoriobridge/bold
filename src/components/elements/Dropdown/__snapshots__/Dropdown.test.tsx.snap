// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should open menu on click 1`] = `
<body>
  .emotion-0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  line-height: 1.5rem;
  position: relative;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-transition: all .2s;
  transition: all .2s;
  font-weight: bold;
  cursor: pointer;
  background-color: #FFFFFF;
  border: 1px solid #818489;
  border-radius: 4px;
  color: #2D343B;
  font-size: 0.875rem;
  padding: calc(0.75rem - 1px) 2.5rem;
}

.emotion-0 > span {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-transition: color .2s;
  transition: color .2s;
}

.emotion-0 > span > :not(:last-child) {
  margin-right: 0.5rem;
}

.emotion-0:not(:disabled):active {
  box-shadow: inset 0 2px 8px 0 rgba(0,0,0,0.1);
}

.emotion-0:focus {
  outline: none;
  box-shadow: 0 0 0 2px #F2F2F7,0 0 0 4px #0A67D1;
}

.emotion-0:not(:disabled):hover {
  background-color: #ebebeb;
}

<div>
    <button
      aria-expanded="true"
      aria-haspopup="true"
      aria-label="close menu"
      aria-labelledby="downshift-1-label"
      aria-owns="downshift-1-menu"
      class="emotion-0"
      data-toggle="true"
      role="button"
      type="button"
    >
      <span>
        <span>
          Menu
        </span>
      </span>
    </button>
  </div>
  .emotion-4 {
  font-weight: bold;
  color: #2D343B;
  white-space: nowrap;
  list-style: none;
  margin: 0;
  padding: 0;
  box-shadow: 0 2px 4px 0 rgba(0,0,0,0.09);
  border: 1px solid #DCDCE0;
  border-radius: 2px;
  display: inline-block;
  width: auto;
  min-width: 150px;
  background: #FFFFFF;
  text-align: left;
}

.emotion-0 {
  display: block;
  padding: 0.5rem 1rem;
  font-size: 0.875rem;
  color: #2D343B;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.emotion-0:hover {
  background: #F2F2F7;
}

.emotion-1 {
  margin: 0;
}

.emotion-1 div {
  display: block;
}

.emotion-1:first-of-type a {
  border-top-left-radius: 2px;
  border-top-right-radius: 2px;
}

.emotion-1:last-of-type a {
  border-bottom-left-radius: 2px;
  border-bottom-right-radius: 2px;
}

.emotion-5 {
  -webkit-transition: opacity 200ms ease-in-out;
  transition: opacity 200ms ease-in-out;
  opacity: 0;
  opacity: 1;
}

<div
    class="emotion-5"
    style="position: absolute; top: 0px; left: 0px; opacity: 0; pointer-events: none; z-index: 1300;"
  >
    <ul
      class="emotion-4"
    >
      <li
        class="emotion-1"
      >
        <a
          class="emotion-0"
        >
          Item #1
        </a>
      </li>
      <li
        class="emotion-1"
      >
        <a
          class="emotion-0"
        >
          Item #2
        </a>
      </li>
    </ul>
  </div>
</body>
`;

exports[`should render correctly 1`] = `
.emotion-0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  line-height: 1.5rem;
  position: relative;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-transition: all .2s;
  transition: all .2s;
  font-weight: bold;
  cursor: pointer;
  background-color: #FFFFFF;
  border: 1px solid #818489;
  border-radius: 4px;
  color: #2D343B;
  font-size: 0.875rem;
  padding: calc(0.75rem - 1px) 2.5rem;
}

.emotion-0 > span {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-transition: color .2s;
  transition: color .2s;
}

.emotion-0 > span > :not(:last-child) {
  margin-right: 0.5rem;
}

.emotion-0:not(:disabled):active {
  box-shadow: inset 0 2px 8px 0 rgba(0,0,0,0.1);
}

.emotion-0:focus {
  outline: none;
  box-shadow: 0 0 0 2px #F2F2F7,0 0 0 4px #0A67D1;
}

.emotion-0:not(:disabled):hover {
  background-color: #ebebeb;
}

<body>
  <div>
    <button
      aria-expanded="false"
      aria-haspopup="true"
      aria-label="open menu"
      aria-labelledby="downshift-0-label"
      class="emotion-0"
      data-toggle="true"
      role="button"
      type="button"
    >
      <span>
        <span>
          Menu
        </span>
      </span>
    </button>
  </div>
</body>
`;
